### ISSUE TITLE
Floating point exception in TCanvas

### ISSUE THREAD

**Issue by andriish** [root-project/root#6379]
## Describe the bug
Crash.
```
[host@hermes Desktop]$ g++ canvas.cxx $(root-config --glibs --cflags --libs) -o canvas  -g
[host@hermes Desktop]$ ./canvas

 *** Break *** floating point exception



===========================================================
There was a crash.
This is the entire stack trace of all threads:
===========================================================
    gdb.printing.register_pretty_printer(gdb.current_objfile(),
#0  0x00007f481d5d946c in __libc_waitpid (pid=1478152, stat_loc=stat_loc
entry=0x7ffe12925160, options=options
entry=0) at ../sysdeps/unix/sysv/linux/waitpid.c:31
#1  0x00007f481d556f62 in do_system (line=<optimized out>) at ../sysdeps/posix/system.c:148
#2  0x00007f4821f685dc in TUnixSystem::StackTrace (this=0x244d980) at /usr/src/debug/root-6.22.02/core/unix/src/TUnixSystem.cxx:2408
#3  0x00007f4821f6b06a in TUnixSystem::DispatchSignals (this=0x244d980, sig=kSigFloatingException) at /usr/src/debug/root-6.22.02/core/unix/src/TUnixSystem.cxx:3646
#4  <signal handler called>
#5  TPad::ResizePad (this=0x2503ca0, option=0x7f4820490bae "") at /usr/src/debug/root-6.22.02/graf2d/gpad/src/TPad.cxx:5490
#6  0x00007f4820448f84 in TCanvas::Build (this=0x2503ca0) at /usr/src/debug/root-6.22.02/graf2d/gpad/src/TCanvas.cxx:637
#7  0x00007f482044969e in TCanvas::Constructor (this=this
entry=0x2503ca0, name=name
entry=0x400cc3 "canvas", title=title
entry=0x400cc3 "canvas", ww=4, ww
entry=-4, wh=wh
entry=-28) at /usr/src/debug/root-6.22.02/graf2d/gpad/src/TCanvas.cxx:425
#8  0x00007f482044c88c in TCanvas::TCanvas (this=0x2503ca0, name=0x400cc3 "canvas", title=0x400cc3 "canvas", ww=-4, wh=-28) at /usr/src/debug/root-6.22.02/graf2d/gpad/src/TCanvas.cxx:366
#9  0x0000000000400ba3 in main (argc=1, argv=0x7ffe12927dc8) at canvas.cxx:5
===========================================================


The lines below might hint at the cause of the crash.
You may get help by asking at the ROOT forum http://root.cern.ch/forum
Only if you are really convinced it is a bug in ROOT then please submit a
report at http://root.cern.ch/bugs Please post the ENTIRE stack trace
from above as an attachment in addition to anything else
that might help us fixing this issue.
===========================================================
#5  TPad::ResizePad (this=0x2503ca0, option=0x7f4820490bae "") at /usr/src/debug/root-6.22.02/graf2d/gpad/src/TPad.cxx:5490
#6  0x00007f4820448f84 in TCanvas::Build (this=0x2503ca0) at /usr/src/debug/root-6.22.02/graf2d/gpad/src/TCanvas.cxx:637
#7  0x00007f482044969e in TCanvas::Constructor (this=this
entry=0x2503ca0, name=name
entry=0x400cc3 "canvas", title=title
entry=0x400cc3 "canvas", ww=4, ww
entry=-4, wh=wh
entry=-28) at /usr/src/debug/root-6.22.02/graf2d/gpad/src/TCanvas.cxx:425
#8  0x00007f482044c88c in TCanvas::TCanvas (this=0x2503ca0, name=0x400cc3 "canvas", title=0x400cc3 "canvas", ww=-4, wh=-28) at /usr/src/debug/root-6.22.02/graf2d/gpad/src/TCanvas.cxx:366
#9  0x0000000000400ba3 in main (argc=1, argv=0x7ffe12927dc8) at canvas.cxx:5
===========================================================

```

## Expected behavior
No FPE.

## To Reproduce
Code
```
#include <TCanvas.h>
#include <fenv.h>
int main(int argc, char **argv) {
    feenableexcept(FE_DIVBYZERO | FE_INVALID | FE_OVERFLOW);
    TCanvas* C= new TCanvas("canvas","canvas",-4,-28);
    return 0;
}

```
Compile
```
g++ canvas.cxx $(root-config --glibs --cflags --libs) -o canvas  -g
```
run

```
./canvas
```

## Setup
CentOS7 x86_64 with root-6.22.02 from EPEL
(Same as on lxplus)

## Additional context

This is a continuation of  #6344

**Comment by couet** [comment]
I get the FPE on ubuntu too.

**Comment by couet** [comment]
I see the crash on Ubuntu with that one:
```
$ ./canvas

 *** Break *** floating point exception



===========================================================
There was a crash.
This is the entire stack trace of all threads:
===========================================================
#0  0x00007f99355bbc2a in __GI___wait4 (pid=474567, stat_loc=stat_loc
entry=0x7ffe7e74a968, options=options
entry=0, usage=usage
entry=0x0) at ../sysdeps/unix/sysv/linux/wait4.c:27
#1  0x00007f99355bbbeb in __GI___waitpid (pid=<optimized out>, stat_loc=stat_loc
entry=0x7ffe7e74a968, options=options
entry=0) at waitpid.c:38
#2  0x00007f993552b0e7 in do_system (line=<optimized out>) at ../sysdeps/posix/system.c:172
#3  0x00007f9935e0f82e in TUnixSystem::StackTrace() () from /home/sftnight/couet/build/lib/libCore.so
#4  0x00007f9935e0c6b5 in TUnixSystem::DispatchSignals(ESignals) () from /home/sftnight/couet/build/lib/libCore.so
#5  <signal handler called>
#6  0x00007f9935ab4ad0 in TPad::ResizePad(char const*) () from /home/sftnight/couet/build/lib/libGpad.so
#7  0x00007f9935a7dc68 in TCanvas::Build() () from /home/sftnight/couet/build/lib/libGpad.so
#8  0x00007f9935a7e5a4 in TCanvas::Constructor(char const*, char const*, int, int) () from /home/sftnight/couet/build/lib/libGpad.so
#9  0x00007f9935a809c1 in TCanvas::TCanvas(char const*, char const*, int, int) () from /home/sftnight/couet/build/lib/libGpad.so
#10 0x0000563a3676a237 in main (argc=1, argv=0x7ffe7e74d7c8) at canvas.cxx:5
===========================================================


The lines below might hint at the cause of the crash.
You may get help by asking at the ROOT forum http://root.cern.ch/forum
Only if you are really convinced it is a bug in ROOT then please submit a
report at http://root.cern.ch/bugs Please post the ENTIRE stack trace
from above as an attachment in addition to anything else
that might help us fixing this issue.
===========================================================
#6  0x00007f9935ab4ad0 in TPad::ResizePad(char const*) () from /home/sftnight/couet/build/lib/libGpad.so
#7  0x00007f9935a7dc68 in TCanvas::Build() () from /home/sftnight/couet/build/lib/libGpad.so
#8  0x00007f9935a7e5a4 in TCanvas::Constructor(char const*, char const*, int, int) () from /home/sftnight/couet/build/lib/libGpad.so
#9  0x00007f9935a809c1 in TCanvas::TCanvas(char const*, char const*, int, int) () from /home/sftnight/couet/build/lib/libGpad.so
#10 0x0000563a3676a237 in main (argc=1, argv=0x7ffe7e74d7c8) at canvas.cxx:5
===========================================================

```

**Comment by couet** [comment]
A protection against negative canvas height should be added. I will make a PR.

**Comment by couet** [comment]
I see the problem also on Mac with interactive ROOT. 
```
root [0]     TCanvas* C= new TCanvas("canvas","canvas",-4,-28);

2020-09-22 11:25:25.425 root.exe[23095:6904258] QuartzPixmap: -resizeW:H:, CGBitmapContextCreateWithData failed
Error in <TGQuartz::SetDoubleBufferON>: QuartzPixmap initialization failed
Warning in <TCanvas::ResizePad>: canvas height changed from 32000 to 10

2020-09-22 11:25:25.742 root.exe[23095:6904258] *** Assertion failure in -[_NSCGSWindow setFrame:], /AppleInternal/BuildRoot/Library/Caches/com.apple.xbs/Sources/AppKit/AppKit-1894.60.100/CGS.subproj/NSCGSWindow.m:1010
2020-09-22 11:25:25.744 root.exe[23095:6904258] *** Terminating app due to uncaught exception 'NSInternalInconsistencyException', reason: 'Invalid parameter not satisfying: CGRectContainsRect(CGRectMake((CGFloat)INT_MIN, (CGFloat)INT_MIN, (CGFloat)INT_MAX - (CGFloat)INT_MIN, (CGFloat)INT_MAX - (CGFloat)INT_MIN), frame)'
*** First throw call stack:
(
	0   CoreFoundation                      0x00007fff36641b57 __exceptionPreprocess + 250
	1   libobjc.A.dylib                     0x00007fff6f4885bf objc_exception_throw + 48
	2   CoreFoundation                      0x00007fff3666ad08 +[NSException raise:format:arguments:] + 88
	3   Foundation                          0x00007fff38d5ce9d -[NSAssertionHandler handleFailureInMethod:object:file:lineNumber:description:] + 191
	4   AppKit                              0x00007fff33988b55 -[_NSCGSWindow setFrame:] + 428
	5   AppKit                              0x00007fff33987add _NSCreateWindowWithOpaqueShape2 + 241
	6   AppKit                              0x00007fff33986920 -[NSWindow _commonAwake] + 1121
	7   AppKit                              0x00007fff339864bb NSWindowWakeDeferredIfNeeded + 94
	8   AppKit                              0x00007fff3398637d -[NSWindow(NSFullScreen) _resolveAutomaticEnterFullScreenFlags] + 22
	9   AppKit                              0x00007fff33986333 -[NSWindow makeKeyAndOrderFront:] + 27
	10  libGCocoa.so                        0x000000011a10369b -[QuartzWindow makeKeyAndOrderFront:] + 75
	11  libGCocoa.so                        0x000000011a1042dd -[QuartzWindow mapRaised] + 77
	12  libGCocoa.so                        0x000000011a10bf44 _ZN7TGCocoa9MapRaisedEm + 196
	13  libGui.so                           0x000000011a196fee _ZN7TGFrame9MapRaisedEv + 14
	14  libGpad.so                          0x0000000115b40b92 _ZN7TCanvas11ConstructorEPKcS1_ii + 610
	15  libGpad.so                          0x0000000115b408dc _ZN7TCanvasC2EPKcS1_ii + 220
	16  ???                                 0x0000000115b38064 0x0 + 4659052644
	17  ???                                 0x0000000115b380c0 0x0 + 4659052736
	18  libCling.so                         0x0000000104884a81 _ZNK5cling19IncrementalExecutor25runStaticInitializersOnceERKNS_11TransactionE + 625
	19  libCling.so                         0x00000001048989aa _ZN5cling11Interpreter18executeTransactionERNS_11TransactionE + 154
	20  libCling.so                         0x000000010488ef72 _ZN5cling17IncrementalParser17commitTransactionERN4llvm14PointerIntPairIPNS_11TransactionELj2ENS0_12EParseResultENS1_21PointerLikeTypeTraitsIS4_EENS1_18PointerIntPairInfoIS4_Lj2ES7_EEEEb + 834
	21  libCling.so                         0x000000010488f8a1 _ZN5cling17IncrementalParser7CompileEN4llvm9StringRefERKNS_18CompilationOptionsE + 289
	22  libCling.so                         0x0000000104897650 _ZN5cling11Interpreter16EvaluateInternalERKNSt3__112basic_stringIcNS1_11char_traitsIcEENS1_9allocatorIcEEEENS_18CompilationOptionsEPNS_5ValueEPPNS_11TransactionEm + 160
	23  libCling.so                         0x0000000104896ee3 _ZN5cling11Interpreter7processERKNSt3__112basic_stringIcNS1_11char_traitsIcEENS1_9allocatorIcEEEEPNS_5ValueEPPNS_11TransactionEb + 355
	24  libCling.so                         0x00000001048d3ceb _ZN5cling13MetaProcessor7processEN4llvm9StringRefERNS_11Interpreter17CompilationResultEPNS_5ValueEb + 331
	25  libCling.so                         0x000000010478bbff _ZL26HandleInterpreterExceptionPN5cling13MetaProcessorEPKcRNS_11Interpreter17CompilationResultEPNS_5ValueE + 63
	26  libCling.so                         0x000000010478b7da _ZN6TCling11ProcessLineEPKcPN12TInterpreter10EErrorCodeE + 3498
	27  libRint.so                          0x0000000103d981d6 _ZN5TRint13ProcessLineNrEPKcS1_Pi + 198
	28  libRint.so                          0x0000000103d98d54 _ZN5TRint15HandleTermInputEv + 564
	29  libCore.so                          0x0000000103f17b22 _ZN11TUnixSystem16CheckDescriptorsEv + 274
	30  libCore.so                          0x0000000103f14060 _ZN13TMacOSXSystem16DispatchOneEventEb + 384
	31  libCore.so                          0x0000000103e3c79a _ZN7TSystem9InnerLoopEv + 26
	32  libCore.so                          0x0000000103e3c662 _ZN7TSystem3RunEv + 306
	33  libCore.so                          0x0000000103ddd214 _ZN12TApplication3RunEb + 36
	34  libRint.so                          0x0000000103d9805a _ZN5TRint3RunEb + 1402
	35  root.exe                            0x0000000103d87edf main + 79
	36  libdyld.dylib                       0x00007fff70630cc9 start + 1
	37  ???                                 0x0000000000000002 0x0 + 2
)
libc++abi.dylib: terminating with uncaught exception of type NSException
$
```

**Comment by couet** [comment]
Fixed by #6424
